Issue 405 document cellar vs fsbuckets (#407)
CleverCloud/doc.clever-cloud.com: commons/addons/cellar-vs-fsbuckets.md

---
title: Choosing between File System Buckets and Cellar S3
position: 3

> shortdesc: This will help you understand the differenc between Cellar and FS Buckets.

Reviewer: difference*

tags:
- addons
keywords:
- S3
- amazon
- Storage
- file
- files
---

You want to store files but you can't decide between FS Buckets and Cellar. This article

> will compare them: usage, pros/cons, pricing. Get ready for the ultimate comparison.

Reviewer 2: > ultimate

avoid superlative in documentation, it's not marketing

Author: Fixed. Removed the "sensational" sentence ;)

> Clever Cloud uses immutable throwable VMs.

Reviewer 2: s/throwable/disposable/g

Author: Thanks, at the time I could not find the right word… disposable is exactly was I was looking for.

> Every time you redeploy your application, you lose the old VMs and all the files stored on their filesystem.

Reviewer 3: ```suggestion
Every time you redeploy your application, you lose the old VM and all the files stored on their filesystem.
```

> If you want to avoid that, you might want to store your important files outside of your instances.

Reviewer 2: you might want to -> you have to

Cellar and FS Buckets both allow you to store files ouside of your instances for later
use. But there are some differences between them.

<div class="panel panel-warning">
  <div class="panel-heading">
    <h4 class="panel-title">Don't want to read the guide?</h4>
  </div>
  <div class="panel-body">
    If you are coding a new project and need to store files, use Cellar. It will be
    cheaper and give you way more flexibility.<br />

>     If your want to run a Wordpress site or some OSS PHP application and are in a hurry,

Reviewer 3: ```suggestion
    If your want to run a WordPress site or some OSS PHP application and are in a hurry,
```

    you may not get to choose: use FS Buckets.
  </div>
</div>

## What is FS Bucket?

Configuring FS Buckets in your application will give you a folder _in your filesystem_.
This folder is shared between instances/applications. It's mounted over the network.
As for any network filesystem there are upsides and downsides.

### Upsides

If your application manages files on a filesystem, you don't have to change anything.
Just setup your FS Bucket to mount on a given folder, setup your application to write in that folder and you are good to go!

Since the folder is mounted over the network, multiple instances/applications can read it
like any file on the system. At the same time.

### Downsides

> Since it's a file on a filesystem, you can be tempted to store a sqlite (or equivalent) database in it.

Reviewer 3: ```suggestion
Since it's a file on a filesystem, you can be tempted to store a SQLite (or equivalent) database in it.
```

This would be wrong.
If many instances try to write in it at the same time, you might end up with a corrupted
database.

For the same reason, anything accessing a file in write at the same time might corrupt the
data you put in it.

## What is Cellar?

Cellar is an Amazon S3 API implementation using [Ceph Rados Gateway](https://ceph.com/ceph-storage/object-storage/).
It provides an HTTP API that can be used with any S3 API Client.

> Clients exists for a lot of languages.

Reviewer 3: ```suggestion
Clients exist for a lot of languages.
```

You can use any S3 Client, from the official Amazon ones to the community ones.

> If you already use S3, you will be in your element.

Reviewer 3: ```suggestion
If you already using S3, you will be in your element.
```

To store a file, issue a PUT HTTP request to Cellar's API.
To get a file, issue a GET request to the API.
You need secret and public keys to authenticate your requests.

> You can manage ACLs, generate signed URLs with an expiration date.

Reviewer 3: ```suggestion
You can manage ACLs, and generate signed URLs with an expiration date.
```

### Upsides

You can move the file uploads to your client entirely.

> Just give it a temporary signed URL and upload directly from the client to Cellar.

Reviewer 3: ```suggestion
Just give it a temporary signed URL (generated by your backend) and it will be able to upload directly to Cellar.
```

That way, you don't have to handle the receive-and-upload part on the server side.

Your files are stored in a secure way outside of your apps FS. So an attacker will have a
harder time to find it.

> You can delegate the authorization part to Cellar by giving signed urls for file download

Reviewer 3: ```suggestion
You can delegate the authorization part to Cellar by giving signed urls for file downloads
```

and uploads.

The files are stored with a replication factor of 2 or 3, making a hardware drive failure
less critical.

### Downsides

When deploying an application you did not write, file uploads may already be developed

> using local files. Like in Wordpress, Prestashop, etc. So you would have to use plugins or

Reviewer 3: ```suggestion
using local files. Like in WordPress, PrestaShop, etc. So you would have to use plugins or
```

hack it to upload to Cellar.

## What should I choose?

Well, you read the upsides and downsides of both.
You may understand where our preference goes, by now!

> I you can, use Cellar.

Reviewer 3: ```suggestion
If you can, use Cellar.
```

will compare them: usage, pros/cons, pricing. Get ready for the ultimate comparison.

Clever Cloud uses immutable throwable VMs.

> Every time you redeploy your application, you lose the old VM and all the files stored on their filesystem.

Author: Actually, if you remove the "s", change "their" by "its".

Reviewer 3: ```suggestion
Every time you redeploy your application, you lose the old VMs and all the files stored on their filesystems.
```

Author: I actually changed to "instances".

Cellar is an Amazon S3 API implementation using [Ceph Rados Gateway](https://ceph.com/ceph-storage/object-storage/).
It provides an HTTP API that can be used with any S3 API Client.
Clients exist for a lot of languages.
You can use any S3 Client, from the official Amazon ones to the community ones.

> If you already using S3, you will be in your element.

Author: Nope ^^
Read the sentence again ;)

Reviewer 3: ```suggestion
If you are already using S3, you will be in your element.
```

Author: I'll keep my version. It's still correct and it's shorter ;)

To store a file, issue a PUT HTTP request to Cellar's API.
To get a file, issue a GET request to the API.
You need secret and public keys to authenticate your requests.

> You can manage ACLs, and generate signed URLs with an expiration date.

Author: Remove comma if you add "and".

Reviewer 3: ```suggestion
You can manage ACLs and you can generate signed URLs with an expiration date.
```
It is a bit repetitive but it is better to separate the two independents ideas, isn't it?
